spectral_cube/io/casa_low_level_io.py

```diff
@@ -0,0 +1,447 @@
+        if end - start != nbytes:
+            raise IOError('Function {0} read {1} bytes instead of {2}'
+                          .format(func, end - start, nbytes))
+        return result
+    return wrapper
+
+
+def read_bool(f):
+    return f.read(1) == b'\x01'
+
+
+def read_int32(f):
+    return np.int32(struct.unpack('>I', f.read(4))[0])
+
+
+def read_int64(f):
+    return np.int64(struct.unpack('>Q', f.read(8))[0])
+
+
+def read_float32(f):
+    return np.float32(struct.unpack('>f', f.read(4))[0])
+
+
+def read_float64(f):
+    return np.float64(struct.unpack('>d', f.read(8))[0])
+
+
+def read_complex128(f):
+    return np.complex128(read_float64(f) + 1j * read_float64(f))
+
+
+def read_complex64(f):
+    return np.complex64(read_float32(f) + 1j * read_float32(f))
+
+
+def read_string(f):
+    value = read_int32(f)
+    return f.read(int(value)).decode('ascii')
+
+
+@with_nbytes_prefix
+def read_iposition(f):
+
+    stype, sversion = read_type(f)
+
+    if stype != 'IPosition' or sversion != 1:
+        raise NotImplementedError('Support for {0} version {1} not implemented'.format(stype, sversion))
+
+    nelem = read_int32(f)
+
+    return np.array([read_int32(f) for i in range(nelem)])
+
+
+ARRAY_ITEM_READERS = {
+    'float': ('float', read_float32, np.float32),
+    'double': ('double', read_float64, np.float64),
+    'dcomplex': ('void', read_complex128, np.complex128),
+    'string': ('String', read_string, '<U16'),
+    'int': ('Int', read_int32, np.int32)
+}
+
+
+@with_nbytes_prefix
+def read_array(f, arraytype):
+
+    typerepr, reader, dtype = ARRAY_ITEM_READERS[arraytype]
+
+    stype, sversion = read_type(f)
+
+    if stype != f'Array<{typerepr}>' or sversion != 3:
+        raise NotImplementedError('Support for {0} version {1} not implemented'.format(stype, sversion))
+
+    ndim = read_int32(f)
+    shape = [read_int32(f) for i in range(ndim)]
+    size = read_int32(f)
+
+    values = [reader(f) for i in range(size)]
+
+    return np.array(values, dtype=dtype).reshape(shape)
+
+
+def read_type(f):
+    tp = read_string(f)
+    version = read_int32(f)
+    return tp, version
+
+
+@with_nbytes_prefix
+def read_record(f):
+
+    stype, sversion = read_type(f)
+
+    if stype != 'Record' or sversion != 1:
+        raise NotImplementedError('Support for {0} version {1} not implemented'.format(stype, sversion))
+
+    read_record_desc(f)
+
+    # Not sure what the following value is
+    read_int32(f)
+
+
+@with_nbytes_prefix
+def read_record_desc(f):
+
+    stype, sversion = read_type(f)
+
+    if stype != 'RecordDesc' or sversion != 2:
+        raise NotImplementedError('Support for {0} version {1} not implemented'.format(stype, sversion))
+
+    # Not sure what the following value is
+    nrec = read_int32(f)
+
+    records = OrderedDict()
+
+    for i in range(nrec):
+        name = read_string(f)
+        rectype = TYPES[read_int32(f)]
+        records[name] = {'type': rectype}
+        if rectype == 'bool':
```

Comment:
not sure if it's worth it, but some of this could be shrunk down by doing 
`if rectype in ('bool', 'int, 'uint'): records[name]['value'] = read_int32(f)` and similar

Reply:
Done